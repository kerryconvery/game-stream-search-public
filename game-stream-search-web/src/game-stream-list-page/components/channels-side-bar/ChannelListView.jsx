import React from 'react';
import { shape, string, bool, arrayOf, func } from 'prop-types';
import { makeStyles, styled } from '@material-ui/core/styles';
import Avatar from '@material-ui/core/Avatar';
import Skeleton from '@material-ui/lab/Skeleton';
import Link from '../../../shared-components/Link';

const useChannelTileStyles = makeStyles({
  hover: {
    '&:hover > *': {
      backgroundColor: '#E4E3E3',
    }
  },
  channelTile: {
    display: 'flex',
    flexDirection: 'row',
    padding: '0.5rem',
    paddingLeft: '1rem',
  },
  channelDetails: {
    paddingTop: '0.25rem',
    paddingLeft: '0.5rem',
    '& > div': {
      paddingBottom: '0.2rem'
    }
  },
})

const StreamPlatformName = styled('span')({
  color: '#606060',
  fontSize: '14px',
});

const ChannelName = styled('div')({
  fontWeight: 'bold'
})

const ChannelTile = ({ channelName , streamPlatformDisplayName, avatarUrl, channelUrl, onChannelOpened }) => {
  const classes = useChannelTileStyles();

  return (
    <Link href={channelUrl} onClick={onChannelOpened} target='_blank'>
      <div className={classes.hover}>
        <div className={classes.channelTile}>
          <Avatar src={avatarUrl} />
          <div className={classes.channelDetails}>
            <ChannelName>{channelName}</ChannelName>
            <StreamPlatformName>{streamPlatformDisplayName}</StreamPlatformName>
          </div>
        </div>
      </div>
    </Link>
  )
}

ChannelTile.propTypes = {
  channel: shape({
    channelName: string.isRequired,
    streamPlatformDisplayName: string.isRequired,
    avatarUrl: string.isRequired,
    channelUrl: string.isRequired,
  }),
  onChannelOpened: func,
}

const LoadingTile = () => {
  const classes = useChannelTileStyles();

  return (
    <div data-testid='channel-loading-tile'>
      <div className={classes.channelTile}>
        <Skeleton variant='circle' width={50} height={50} animation='wave' />
        <div className={classes.channelDetails}>
          <Skeleton variant='text' width={120} animation='wave' />
          <Skeleton variant='text' width={80} animation='wave' />
        </div>
      </div>
    </div>
  )
}

const useChannelListStyles = makeStyles({
  channelList: {
    display: 'grid',
    gridTemplateColumns: 'auto',
    gridAutoFlow: 'row',
  },
})

export const ChannelListView = ({ channels, isLoading, onChannelOpened }) => {
  const classes = useChannelListStyles();

  const channelTitles = channels.map((channel, index) => (
    <ChannelTile key={index} {...channel} onChannelOpened={() => onChannelOpened(channel)} />
  ));

  const loadingTiles = [];

  if (isLoading)
  {
    for (let index = 0; index < 3; index++) {
      loadingTiles.push(<LoadingTile key={index} />);
    }
  }

  return (
    <div className={classes.channelList}>
      {channelTitles}
      {loadingTiles}
    </div>
  )
}

ChannelListView.propTypes = {
  channels: arrayOf(shape({
    channelName: string.isRequired,
    streamPlatformDisplayName: string.isRequired,
    avatarUrl: string.isRequired,
    channelUrl: string.isRequired,
  })),
  isLoading: bool,
  onChannelOpened: func,
};

ChannelListView.defaultProps = {
  streams: [],
  isLoading: false,
};

export default ChannelListView;