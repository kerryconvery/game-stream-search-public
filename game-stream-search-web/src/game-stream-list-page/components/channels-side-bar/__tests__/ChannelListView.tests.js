import React from 'react';
import { render } from '@testing-library/react';
import ChannelListView from '../ChannelListView';

describe('Channel List', () => {
  it('should render a list of channels without loading tiles', () => {
    const channels = [
      {
        channelName: 'test channel 1',
        streamPlatformDisplayName: 'YouTube',
        avatarUrl: '',
        channelUrl: '',
      },
      {
        channelName: 'test channel 2',
        streamPlatformDisplayName: 'DLive',
        avatarUrl: '',
        channelUrl: '',
      }
    ];

    const { container } = render(<ChannelListView channels={channels} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a list of loading tiles', () => {
    const { container } = render(<ChannelListView channels={[]} isLoading />);

    expect(container.firstChild).toMatchSnapshot();
  })
})