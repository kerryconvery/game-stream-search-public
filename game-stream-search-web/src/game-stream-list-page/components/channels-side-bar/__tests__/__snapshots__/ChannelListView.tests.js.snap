// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel List should render a list of channels without loading tiles 1`] = `
<div
  class="makeStyles-channelList-1"
>
  <a
    class="a-root-5"
    href=""
    target="_blank"
  >
    <div
      class="makeStyles-hover-2"
    >
      <div
        class="makeStyles-channelTile-3"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiAvatar-fallback"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
        <div
          class="makeStyles-channelDetails-4"
        >
          <div
            class="div-root-6"
          >
            test channel 1
          </div>
          <span
            class="span-root-7"
          >
            YouTube
          </span>
        </div>
      </div>
    </div>
  </a>
  <a
    class="a-root-5"
    href=""
    target="_blank"
  >
    <div
      class="makeStyles-hover-2"
    >
      <div
        class="makeStyles-channelTile-3"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiAvatar-fallback"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
        <div
          class="makeStyles-channelDetails-4"
        >
          <div
            class="div-root-6"
          >
            test channel 2
          </div>
          <span
            class="span-root-7"
          >
            DLive
          </span>
        </div>
      </div>
    </div>
  </a>
</div>
`;

exports[`Channel List should render a list of loading tiles 1`] = `
<div
  class="makeStyles-channelList-8"
>
  <div
    data-testid="channel-loading-tile"
  >
    <div
      class="makeStyles-channelTile-10"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-wave"
        style="width: 50px; height: 50px;"
      />
      <div
        class="makeStyles-channelDetails-11"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 120px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 80px;"
        />
      </div>
    </div>
  </div>
  <div
    data-testid="channel-loading-tile"
  >
    <div
      class="makeStyles-channelTile-10"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-wave"
        style="width: 50px; height: 50px;"
      />
      <div
        class="makeStyles-channelDetails-11"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 120px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 80px;"
        />
      </div>
    </div>
  </div>
  <div
    data-testid="channel-loading-tile"
  >
    <div
      class="makeStyles-channelTile-10"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-wave"
        style="width: 50px; height: 50px;"
      />
      <div
        class="makeStyles-channelDetails-11"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 120px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          style="width: 80px;"
        />
      </div>
    </div>
  </div>
</div>
`;
